core/storage/pager.rs

```diff
@@ -347,24 +347,23 @@ impl Pager {
         self.page_cache.borrow_mut().resize(capacity);
     }
 
-    pub fn add_dirty(&self, page: Rc<RefCell<Page>>) {
+    pub fn add_dirty(&self, page_id: usize) {
         // TODO: cehck duplicates?
         let mut dirty_pages = RefCell::borrow_mut(&self.dirty_pages);
-        dirty_pages.push(page);
+        dirty_pages.insert(page_id);
     }
 
     pub fn cacheflush(&self) -> Result<()> {
         let mut dirty_pages = RefCell::borrow_mut(&self.dirty_pages);
         if dirty_pages.len() == 0 {
             return Ok(());
         }
-        loop {
-            if dirty_pages.len() == 0 {
-                break;
-            }
-            let page = dirty_pages.pop().unwrap();
+        for page_id in dirty_pages.iter() {
+            let mut cache = self.page_cache.borrow_mut();
+            let page = cache.get(&page_id).expect("we somehow added a page to dirty list but we didn't mark it as dirty, causing cache to drop it.");
             sqlite3_ondisk::begin_write_btree_page(self, &page)?;
```

Comment:
What's the thinking here in terms of failures? We leave the page to `dirty_pages` set and it's fine to just flush it out again later?

Reply:
There is no plan for failures now as I expect changes in the future. This is quite rudimentary now. What kind of failure where you thinking about?